src/github.com/matrix-org/dendrite/common/httpapi.go

```diff
@@ -24,20 +38,32 @@ func MakeAuthAPI(
 			return *err
 		}
 
+		// Add user information to opentracing span
+		span := opentracing.SpanFromContext(req.Context())
+		span.SetTag("matrix.user", device.UserID)
+		span.SetTag("matrix.device", device.ID)
+
 		return f(req, device)
 	}
-	return MakeExternalAPI(metricsName, h)
+	return MakeExternalAPI(tracer, metricsName, h)
 }
 
 // MakeExternalAPI turns a util.JSONRequestHandler function into an http.Handler.
 // This is used for APIs that are called from the internet.
-func MakeExternalAPI(metricsName string, f func(*http.Request) util.JSONResponse) http.Handler {
+func MakeExternalAPI(tracer opentracing.Tracer, metricsName string, f func(*http.Request) util.JSONResponse) http.Handler {
 	h := util.MakeJSONAPI(util.NewJSONRequestHandler(f))
 	withSpan := func(w http.ResponseWriter, req *http.Request) {
-		span := opentracing.StartSpan(metricsName)
+		span := tracer.StartSpan(metricsName)
 		defer span.Finish()
+
+		ext.HTTPUrl.Set(span, req.URL.String())
+		ext.HTTPMethod.Set(span, req.Method)
+
 		req = req.WithContext(opentracing.ContextWithSpan(req.Context(), span))
+		rw := newStatusCodeResponseWriter(w)
```

Comment:
I'm a little confused on the few lines above, hence no comments. I get that this is essentially adding metadata to the trace such as request URI and status code. Do we need to do this for internalAPIs as well.

Reply:
Yeah, this is just annotating the span with some info. We probably want to annotate internalAPI spans with if there was an error or not.